File: src/utils/drivers-license-calculator.ts
/**
 * Malta Driver's License Fees Calculator
 * Based on Transport Malta licensing regulations
 */

export interface DriversLicenseInput {
    /** License type */
    licenseType: 'new' | 'renewal' | 'international' | 'replacement' | 'category_upgrade';
    /** Age of applicant */
    age: number;
    /** License category */
    category: 'A' | 'A1' | 'A2' | 'AM' | 'B' | 'BE' | 'C' | 'C1' | 'CE' | 'D' | 'D1' | 'DE';
    /** License validity period in years */
    validityPeriod: 1 | 5 | 10;
    /** Include theory test */
    includeTheoryTest: boolean;
    /** Include practical test */
    includePracticalTest: boolean;
    /** Is it first license */
    isFirstLicense: boolean;
}

export interface DriversLicenseOutput {
    /** Total cost */
    totalCost: number;
    /** License fee */
    licenseFee: number;
    /** Theory test fee */
    theoryTestFee: number;
    /** Practical test fee */
    practicalTestFee: number;
    /** Medical certificate fee (estimated) */
    medicalFee: number;
    /** Photo fee (estimated) */
    photoFee: number;
    /** Description */
    description: string;
    /** Validity period */
    validity: string;
    /** Breakdown */
    breakdown: { label: string; amount: number }[];
}

/**
 * License fees by type (EUR)
 */
const LICENSE_FEES = {
    new: {
        '1': 15,
        '5': 30,
        '10': 45,
    },
    renewal: {
        '1': 15,
        '5': 30,
        '10': 45,
    },
    international: 23,
    replacement: 25,
    category_upgrade: 35,
};

/**
 * Test fees (EUR)
 */
const TEST_FEES = {
    theory: 30,
    practical: {
        AM: 45,  // Moped
        A1: 55,  // Motorcycle up to 125cc
        A2: 55,  // Motorcycle up to 35kW
        A: 55,   // Motorcycle unrestricted
        B: 60,   // Car
        BE: 70,  // Car with trailer
        C1: 80,  // Light lorry
        C: 90,   // Lorry
        CE: 100, // Lorry with trailer
        D1: 90,  // Minibus
        D: 100,  // Bus
        DE: 110, // Bus with trailer
    },
};

/**
 * Additional fees (estimated)
 */
const ADDITIONAL_FEES = {
    medical: 40,  // Medical certificate (doctor visit)
    photo: 8,     // Passport photos
};

/**
 * Category descriptions
 */
const CATEGORY_DESCRIPTIONS: { [key: string]: string } = {
    AM: 'Moped (up to 50cc)',
    A1: 'Motorcycle up to 125cc',
    A2: 'Motorcycle up to 35kW',
    A: 'Full Motorcycle License',
    B: 'Car License',
    BE: 'Car with Trailer',
    C1: 'Light Lorry (3.5-7.5t)',
    C: 'Heavy Lorry (7.5t+)',
    CE: 'Lorry with Trailer',
    D1: 'Minibus (up to 16 passengers)',
    D: 'Bus (17+ passengers)',
    DE: 'Bus with Trailer',
};

/**
 * Validity based on age
 */
function getMaxValidity(age: number): number {
    if (age >= 70) return 1;  // Over 70: max 1 year
    if (age >= 65) return 5;  // 65-70: max 5 years
    return 10;                 // Under 65: max 10 years
}

/**
 * Calculate driver's license fees
 */
export function calculateDriversLicenseFees(input: DriversLicenseInput): DriversLicenseOutput {
    const { licenseType, age, category, validityPeriod, includeTheoryTest, includePracticalTest, isFirstLicense } = input;

    const breakdown: { label: string; amount: number }[] = [];
    const maxValidity = getMaxValidity(age);
    const effectiveValidity = Math.min(validityPeriod, maxValidity);

    // Calculate license fee
    let licenseFee = 0;
    if (licenseType === 'international') {
        licenseFee = LICENSE_FEES.international;
        breakdown.push({ label: 'International License Fee', amount: licenseFee });
    } else if (licenseType === 'replacement') {
        licenseFee = LICENSE_FEES.replacement;
        breakdown.push({ label: 'Replacement License Fee', amount: licenseFee });
    } else if (licenseType === 'category_upgrade') {
        licenseFee = LICENSE_FEES.category_upgrade;
        breakdown.push({ label: 'Category Upgrade Fee', amount: licenseFee });
    } else {
        const fees = LICENSE_FEES[licenseType as 'new' | 'renewal'];
        licenseFee = fees[effectiveValidity.toString() as '1' | '5' | '10'];
        breakdown.push({ label: `License Fee (${effectiveValidity} year${effectiveValidity > 1 ? 's' : ''})`, amount: licenseFee });
    }

    // Theory test fee
    let theoryTestFee = 0;
    if (includeTheoryTest && (isFirstLicense || licenseType === 'category_upgrade')) {
        theoryTestFee = TEST_FEES.theory;
        breakdown.push({ label: 'Theory Test Fee', amount: theoryTestFee });
    }

    // Practical test fee
    let practicalTestFee = 0;
    if (includePracticalTest && (isFirstLicense || licenseType === 'category_upgrade')) {
        practicalTestFee = TEST_FEES.practical[category] || TEST_FEES.practical.B;
        breakdown.push({ label: `Practical Test Fee (${category})`, amount: practicalTestFee });
    }

    // Medical fee (required for first license and renewals over 65)
    let medicalFee = 0;
    if (isFirstLicense || (licenseType === 'renewal' && age >= 65) || ['C', 'C1', 'CE', 'D', 'D1', 'DE'].includes(category)) {
        medicalFee = ADDITIONAL_FEES.medical;
        breakdown.push({ label: 'Medical Certificate (estimated)', amount: medicalFee });
    }

    // Photo fee
    const photoFee = isFirstLicense || licenseType === 'renewal' ? ADDITIONAL_FEES.photo : 0;
    if (photoFee > 0) {
        breakdown.push({ label: 'Passport Photos (estimated)', amount: photoFee });
    }

    // Total cost
    const totalCost = licenseFee + theoryTestFee + practicalTestFee + medicalFee + photoFee;

    // Generate description
    const categoryDesc = CATEGORY_DESCRIPTIONS[category] || category;
    let description = '';
    if (licenseType === 'new' && isFirstLicense) {
        description = `First ${categoryDesc} license application. Valid for ${effectiveValidity} year${effectiveValidity > 1 ? 's' : ''}.`;
    } else if (licenseType === 'renewal') {
        description = `License renewal for ${categoryDesc}. Valid for ${effectiveValidity} year${effectiveValidity > 1 ? 's' : ''}.`;
    } else if (licenseType === 'international') {
        description = 'International Driving Permit valid for 12 months.';
    } else if (licenseType === 'replacement') {
        description = 'Replacement license for lost, stolen, or damaged card.';
    } else {
        description = `Category upgrade to ${categoryDesc}.`;
    }

    if (age >= 65) {
        description += ' Medical certificate required.';
    }

    return {
        totalCost: Math.round(totalCost),
        licenseFee: Math.round(licenseFee),
        theoryTestFee: Math.round(theoryTestFee),
        practicalTestFee: Math.round(practicalTestFee),
        medicalFee: Math.round(medicalFee),
        photoFee: Math.round(photoFee),
        description,
        validity: `${effectiveValidity} year${effectiveValidity > 1 ? 's' : ''}`,
        breakdown,
    };
}

/**
 * Format currency for display
 */
export function formatCurrency(amount: number): string {
    return new Intl.NumberFormat('en-MT', {
        style: 'currency',
        currency: 'EUR',
        minimumFractionDigits: 0,
        maximumFractionDigits: 0,
    }).format(amount);
}

/**
 * Get driver's license info
 */
export function getDriversLicenseInfo() {
    return {
        licenseFees: LICENSE_FEES,
        testFees: TEST_FEES,
        additionalFees: ADDITIONAL_FEES,
        categories: CATEGORY_DESCRIPTIONS,
    };
}

File: ios-app/Packages/CalculationKit/Sources/CalculationKit/Vehicle/DriversLicenseCalculator.swift
import Foundation

/// License application type.
public enum LicenseType: String, CaseIterable, Codable, Sendable {
    case new, renewal, international, replacement, categoryUpgrade = "category_upgrade"
}

/// Driver's license category.
public enum LicenseCategory: String, CaseIterable, Codable, Sendable {
    case am = "AM", a1 = "A1", a2 = "A2", a = "A", b = "B", be = "BE"
    case c = "C", c1 = "C1", ce = "CE", d = "D", d1 = "D1", de = "DE"
}

/// Input for the driver's license calculator.
public struct DriversLicenseInput: Sendable, Codable, Equatable {
    public var licenseType: LicenseType
    public var age: Int
    public var category: LicenseCategory
    public var validityPeriod: Int
    public var includeTheoryTest: Bool
    public var includePracticalTest: Bool
    public var isFirstLicense: Bool

    public init(licenseType: LicenseType, age: Int, category: LicenseCategory,
                validityPeriod: Int, includeTheoryTest: Bool,
                includePracticalTest: Bool, isFirstLicense: Bool) {
        self.licenseType = licenseType; self.age = age; self.category = category
        self.validityPeriod = validityPeriod; self.includeTheoryTest = includeTheoryTest
        self.includePracticalTest = includePracticalTest; self.isFirstLicense = isFirstLicense
    }
}

/// Result of the driver's license fee calculation.
public struct DriversLicenseOutput: Sendable, Codable, Equatable {
    public let totalCost: Money
    public let licenseFee: Money
    public let theoryTestFee: Money
    public let practicalTestFee: Money
    public let medicalFee: Money
    public let photoFee: Money
    public let validity: String
}

/// Malta Driver's License Fees Calculator.
public struct DriversLicenseCalculator: Sendable {
    private static let licenseFees: [String: [String: Money]] = [
        "new": ["1": 15, "5": 30, "10": 45],
        "renewal": ["1": 15, "5": 30, "10": 45],
    ]
    private static let internationalFee: Money = 23
    private static let replacementFee: Money = 25
    private static let categoryUpgradeFee: Money = 35
    private static let theoryTestFee: Money = 30
    private static let practicalTestFees: [LicenseCategory: Money] = [
        .am: 45, .a1: 55, .a2: 55, .a: 55, .b: 60, .be: 70,
        .c1: 80, .c: 90, .ce: 100, .d1: 90, .d: 100, .de: 110,
    ]
    private static let medicalFee: Money = 40
    private static let photoFee: Money = 8

    public init() {}

    private static func getMaxValidity(age: Int) -> Int {
        if age >= 70 { return 1 }
        if age >= 65 { return 5 }
        return 10
    }

    /// Calculates driver's license fees.
    public func calculate(input: DriversLicenseInput) -> DriversLicenseOutput {
        let maxValidity = Self.getMaxValidity(age: input.age)
        let effectiveValidity = Swift.min(input.validityPeriod, maxValidity)

        let licenseFee: Money
        switch input.licenseType {
        case .international: licenseFee = Self.internationalFee
        case .replacement: licenseFee = Self.replacementFee
        case .categoryUpgrade: licenseFee = Self.categoryUpgradeFee
        case .new, .renewal:
            licenseFee = Self.licenseFees[input.licenseType.rawValue]?["\(effectiveValidity)"] ?? 30
        }

        let theoryTestFee: Money
        if input.includeTheoryTest && (input.isFirstLicense || input.licenseType == .categoryUpgrade) {
            theoryTestFee = Self.theoryTestFee
        } else {
            theoryTestFee = 0
        }

        let practicalTestFee: Money
        if input.includePracticalTest && (input.isFirstLicense || input.licenseType == .categoryUpgrade) {
            practicalTestFee = Self.practicalTestFees[input.category] ?? Self.practicalTestFees[.b] ?? 60
        } else {
            practicalTestFee = 0
        }

        let commercialCategories: Set<LicenseCategory> = [.c, .c1, .ce, .d, .d1, .de]
        let medicalFee: Money
        if input.isFirstLicense || (input.licenseType == .renewal && input.age >= 65)
            || commercialCategories.contains(input.category) {
            medicalFee = Self.medicalFee
        } else {
            medicalFee = 0
        }

        let photoFee: Money
        if input.isFirstLicense || input.licenseType == .renewal {
            photoFee = Self.photoFee
        } else {
            photoFee = 0
        }

        let totalCost = licenseFee + theoryTestFee + practicalTestFee + medicalFee + photoFee
        let validityStr = "\(effectiveValidity) year\(effectiveValidity > 1 ? "s" : "")"

        return DriversLicenseOutput(
            totalCost: totalCost.rounded(to: 0),
            licenseFee: licenseFee,
            theoryTestFee: theoryTestFee,
            practicalTestFee: practicalTestFee,
            medicalFee: medicalFee,
            photoFee: photoFee,
            validity: validityStr
        )
    }
}
